eth/fetcher/tx_fetcher.go

```diff
@@ -422,22 +448,28 @@ func (f *TxFetcher) loop() {
 
 					// Stage 2 and 3 share the set of origins per tx
 					if announces := f.announces[ann.origin]; announces != nil {
-						announces[hash] = struct{}{}
+						announces[hash] = ann.metas[i]
 					} else {
-						f.announces[ann.origin] = map[common.Hash]struct{}{hash: {}}
+						f.announces[ann.origin] = map[common.Hash]*txMetadata{hash: ann.metas[i]}
 					}
 					continue
 				}
 				// If the transaction is already known to the fetcher, but not
 				// yet downloading, add the peer as an alternate origin in the
 				// waiting list.
 				if f.waitlist[hash] != nil {
+					// Ignore double announcements from the same peer. This is
+					// especially important if metadata is also passed along to
+					// prevent malicious peers flip-flopping good/bad values.
+					if _, ok := f.waitlist[hash][ann.origin]; ok {
+						continue
+					}
```

Comment:
Just want to double-check the logic here. So, say `tx A` has size `N`. Malicious peer `A` announces it to us, as size `M` and peer `B` announces it as size `N` (correctly).
Now, `peer A` does not deliver, so we instead ask peer `B` for it, and peer `B` delivers a tx with size `N`. 

- Do we hold both meta's, and thus expect `A` to have `M` if coming from `A` and `N` if coming from `B`? 
- If not, how do we prevent an attack where a malicious announcement is used to punish an honest tx-delivery?

Reply:
And also, would it make sense to somehow flag (warn) if we notice this behaviour: peers announcing different meta for a transaction. 
It could point to shenanigans of this sort, and also of implementation flaws.

Reply:
Answering my own question here: 

> * Do we hold both meta's, and thus expect A to have M if coming from A and N if coming from B?

Yes. So attack not possible here

Reply:
... and furthermore, once the honest peer delivers `tx A`, the lie from `peer A` will be discovered and it will be ejected, since we iterate over all announcements in `f.announces`. Good

Reply:
> And also, would it make sense to somehow flag (warn) if we notice this behaviour: peers announcing different meta for a transaction.
It could point to shenanigans of this sort, and also of implementation flaws.

In the end, it will turn into a Warn when the bad peer gets dropped. IMO there's no need to find announce clashes at the moment of announcements, when delivery happens we'll know for sure who's right and if anyone's wrong.

I guess what we could do is to add a second warning log where we log that someone was a "bit" off, but not off enough to drop.